### PR TITLE
Bridge test env

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

Increase test coverage and create a test environment for the bridge that should make it easier to test.
At this point once we have API to sign messages we can reach 100% coverage. All missing tests are related to signature generation, function depending on that, or versioning. API to sign messages and create pub/priv keys are coming that would allow us to create a great testing environment and experience.
Current coverage is as follows
```
+-------------------------+
| Move Coverage Summary   |
+-------------------------+
Module 000000000000000000000000000000000000000000000000000000000000000b::treasury
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::message_types
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::chain_ids
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::message
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::limiter
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::crypto
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::committee
>>> % Module coverage: 100.00
Module 000000000000000000000000000000000000000000000000000000000000000b::bridge
>>> % Module coverage: 51.21
+-------------------------+
| % Move Coverage: 82.32  |
+-------------------------+
```

The test environment needs more work but it will come in time. For now this work was mostly to reach a good test coverage.
It's important to check this in for 2 main reasons: it increases test coverage significantly and allows others to work on it and improve it.

## Test plan 

This is it

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+    };
+
+    //
+    // Token IDs
+    //
+    const BTC_ID: u8 = 1;
+    const ETH_ID: u8 = 2;
+    const USDC_ID: u8 = 3;
+    const USDT_ID: u8 = 4;
+    // const TEST_TOKEN_ID: u8 = 5;
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_setup.move:45]: remove?

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+
+    public struct USDC has drop {}
+
+    public fun create_bridge_token(
+        ctx: &mut TxContext,
+    ): (UpgradeCap, TreasuryCap<USDC>, CoinMetadata<USDC>) {
+        let otw = create_one_time_witness<USDC>();
+        let (treasury_cap, metadata) = create_currency(
+            otw,
+            8,
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_setup.move:645]: 6 to be consistent with production

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+
+    public struct USDT has drop {}
+
+    public fun create_bridge_token(
+        ctx: &mut TxContext,
+    ): (UpgradeCap, TreasuryCap<USDT>, CoinMetadata<USDT>) {
+        let otw = create_one_time_witness<USDT>();
+        let (treasury_cap, metadata) = create_currency(
+            otw,
+            8,
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_setup.move:680]: ditto

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+        env: &mut BridgeEnv,
+        target_chain_id: u8,
+        eth_address: vector<u8>,
+        coin: Coin<T>,
+    ) {
+        let scenario = env.scenario();
+        scenario.next_tx(@0xAAAA);
+        let mut bridge = scenario.take_shared<Bridge>();
+        bridge.send_token(target_chain_id, eth_address, coin, scenario.ctx());
+        test_scenario::return_shared(bridge);
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_setup.move:430]: we can check a few things here:
1. assert the burn happened
2. token_transfer_records has a new entry with correct metadata

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+        treasury_cap: TreasuryCap<T>,
+        upgrade_cap: UpgradeCap,
+        metadata: CoinMetadata<T>,
+        sender: address,
+    ) {
+        let scenario = env.scenario();
+        scenario.next_tx(sender);
+        let mut bridge = scenario.take_shared<Bridge>();
+        bridge.register_foreign_token<T>(treasury_cap, upgrade_cap, &metadata);
+        test_scenario::return_shared(bridge);
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_setup.move:390]: check waiting_room and treasuries have proper changes here

File: crates/sui-framework/packages/bridge/tests/bridge_tests.move [hunk 6]
```diff
@@ -4,373 +4,173 @@
...
-    //         eth_address,
-    //         btc,
-    //         scenario.ctx(),
-    //     );
-    //     test_scenario::return_shared(bridge);
-
-    //     scenario.end();
-    // }
+    #[test]
+    fun test_register_foreign_token() {
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/tests/bridge_tests.move:110]: 1. do we have a test case for `coin::total_supply(&tc) != 0`?
https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/bridge/sources/treasury.move#L103

2. do we have some test cases for `add_new_token`?

**Review by longbowlu — COMMENTED** [review]
overall lgtm!

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move (hunk 1 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/tests/bridge_setup.move:45]: not really, I am going to add functions that return those and it's not worth to remove it for now

File: crates/sui-framework/packages/bridge/tests/bridge_tests.move (hunk 6 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/tests/bridge_tests.move:110]: 1. `test_register_foreign_token_non_zero_supply`
2. `add_default_tokens` to set up the default token does that for the happy path every time you set up the bridge and `test_add_token_price_zero_value`, `test_add_token_malformed_1`, `test_add_token_malformed_2`, `test_add_token_malformed_3` and `test_add_native_token_nop` test the not so happy path

File: crates/sui-framework/packages/bridge/tests/bridge_setup.move (hunk 5 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/tests/bridge_setup.move:471]: done

**Comment by dariorussi** [comment]
thanks @longbowlu for the review and added the `assert` checks into some of the function. That is a good idea.
I'll look into more that we can do there as those are good checks for the environment.
So more coming though at this point I'd rather have this in and iterate.
Anyway take a look

**Review by longbowlu — APPROVED** [review]
thanks!

**Pull request merged**